PCF-ADR-0008-ray-cluster-authentication.md

```diff
@@ -0,0 +1,74 @@
+# Project CodeFlare - RayCluster Authentication
+|                |                                                                                  |
+| -------------- | -------------------------------------------------------------------------------- |
+| Date           | 12/19/2023                                                                       |
+| Scope          |                                                                                  |
+| Status         | provisional                                                                      |
+| Authors        | [Kevin Postlethwait](@KPostOffice)                                               |
+| Supersedes     | N/A                                                                              |
+| Superseded by: | N/A                                                                              |
+| Issues         |                                                                                  |
+| Other docs:    | none                                                                             |
+
+## What
+
+Enabling OpenShift OAuth as an identity provider and authorization mechanism for KubeRay RayClusters.
+
+## Why
+
+Currently there is no authentication by default enabled in front of the Ray Dashboard and we would like to offer a
+unified login experience for the whole of ODH by way of OpenShift OAuth
+
+## Goals
+
+* Protect the dashboard from unauthenticated access by way of an OAuth flow
+* Protect the API endpoint from unauthenticated access by way of bearer tokens
+* Enable HTTPS by default for accessing the RayCluster API and Dashboard
+
+## Non-Goals
+
+* Creating a fully customizable auth(z) solution
+
+## How
+
+In order to solve this we will add a RayCluster reconciler which creates the necessary resources for enabling
```

Comment:
Besides the external resources, there is also the need to inject the oauth-proxy side-car. Is a controller based mutation acceptable for that? That may cause some quick Pod creations / deletions between if the RayCluster is reconciled first by Kuberay and then by the CFO?

Reply:
I was wondering the same thing. The other option is to add the sidecar with a mutating webhook instead.

Reply:
Even if that introduces some complexity, using a mutating webhook may be more appropriate in that case. If we restrict the support to OpenShift, which is a reasonable assumption IMO, we can leverage the service signing certificate service to generate TLS certificates for the webhook.